rustls/src/crypto/hpke.rs

```diff
@@ -0,0 +1,131 @@
+use alloc::boxed::Box;
+use alloc::vec::Vec;
+use core::fmt::Debug;
+
+use crate::msgs::enums::{HpkeAead, HpkeKdf, HpkeKem};
+use crate::Error;
+
+/// A provider for [RFC 9180] Hybrid Public Key Encryption (HPKE).
+///
+/// At a minimum each provider must support base mode and the [HPKE ciphersuite profile] required for
+/// encrypted client hello (ECH):
+///  * KEM: DHKEM(X25519, HKDF-SHA256)
+///  * KDF: HKDF-SHA256
+///  * AEAD: AES-128-GCM
+///
+/// [RFC 9180]: <https://www.rfc-editor.org/rfc/rfc9180.html>
+/// [HPKE ciphersuite profile]: <https://datatracker.ietf.org/doc/html/draft-ietf-tls-esni-17#section-9>
+pub trait HpkeProvider: Debug + Send + Sync + 'static {
+    /// Create an HPKE instance for the given mode, KEM, KDF and AEAD.
+    ///
+    /// May return an error if the mode, KEM, KDF or AEAD are unsupported by the HPKE provider.
+    fn setup(
```

Comment:
A bit of feedback on this design:

- I suggest we reduce this down to precisely what ECH needs, and no more (that means no need to mention mode apart from commentary).
- I suggest avoiding `setup` here, because that has specific use in HPKE -- `open` / `create` / `new`?
- Maybe wrap the parameter choices in a struct? (`HpkeKem` has an implicit KDF, and `HpkeSymmetricCipherSuite` has an explicit one; I think this doesn't need a further KDF specification).

```rust
pub struct HpkeSuite {
  kem: HpkeKem,
  sym: HpkeSymmetricCipherSuite,
}

pub trait HpkeProvider (...) {
  fn open(&self, suite: HpkeSuite) -> Result<Box<dyn Hpke>, Error>;
  fn supports(&self, suite: HpkeSuite) - >bool;
  // or, alternatively,
  fn supported_suites(&self) -> Vec<HpkeSuite>;
}

pub trait Hpke {
    fn setup_sender(
        &self,
        pk_r: &HpkePublicKey,
        info: &[u8],
    ) -> Result<(EncapsulatedSecret, Box<dyn HpkeSender + '_>), Error>;
    // ...
```

I think it's very unlikely that ECH starts requiring Auth or PSK modes (not sure how that would work protocol-wise!), but if that does happen we could make this interface wider (`open_auth` / `open_psk`, etc.) with distinct return traits.

Reply:
> I suggest we reduce this down to precisely what ECH needs, and no more (that means no need to mention mode apart from commentary).

SGTM.

> I suggest avoiding setup here, because that has specific use in HPKE -- open / create / new?

:+1: I avoided `new` because Clippy wants `new` to not take `&self`. Open seems reasonable.

> Maybe wrap the parameter choices in a struct? 

I'll give that a shot.

Reply:
I've implemented the suggested changes in the latest push.